predicators/approaches/initialized_pg3_approach.py

```diff
@@ -135,19 +136,88 @@ def base_nsrt_to_variable_analogy(
             old_to_new_var[old_v] = new_v
         return old_to_new_var
 
+    def is_consistent_with(self, other: _Analogy) -> bool:
+        """Checks whether a given analogy is consistent with a given
+        analogy."""
+        return all(
+            self.predicates[p] == other.predicates[p]
+            for p in self.predicates if p in other.predicates) and \
+            all(
+                self.nsrts[n] == other.nsrts[n]
+                for n in self.nsrts if n in other.nsrts)
+
+    def extend_with(self, other: _Analogy) -> _Analogy:
+        """Creates a new analogy by exteding this one with another analogy.
+
+        The method does not check if the analogies are consistent.
+        """
+        return _Analogy(predicates={
+            **self.predicates,
+            **other.predicates
+        },
+                        nsrts={
+                            **self.nsrts,
+                            **other.nsrts
+                        },
+                        nsrt_variables={
+                            **self.nsrt_variables,
+                            **other.nsrt_variables
+                        })
+
 
 def _find_env_analogies(base_env: BaseEnv, target_env: BaseEnv,
                         base_nsrts: Set[NSRT],
                         target_nsrts: Set[NSRT]) -> List[_Analogy]:
+    """Multishot analogical matching intended to overcome 1-to-1 mapping
+    constraints.
+
+    It follows a simple procedure and focuses on mapping actions:
+        - it does one step of SME
+        - if there are unmatched actions:
+            - remove all *target* actions that are matched
+            - do one SME run again
+            - check if the new mapping is consistent with the previous one
+    """
     # Use external SME module to find analogies.
     smepy.declare_nary("and")
     base_sme_struct = _create_sme_inputs(base_env, base_nsrts)
     target_sme_struct = _create_sme_inputs(target_env, target_nsrts)
+
     analogies: List[_Analogy] = []
+
     for sme_mapping in _query_sme(base_sme_struct, target_sme_struct):
         analogy = _sme_mapping_to_analogy(sme_mapping, base_env, target_env,
                                           base_nsrts, target_nsrts)
-        analogies.append(analogy)
+
+        # Check which target nsrts are missing.
+        missing_nsrts = target_nsrts - {n for _, n in analogy.nsrts.items()}
+
```

Comment:
so this all will happen at most one time. do you anticipate that we would want to keep trying to extend the analogy later on? or should one extension be enough in general?

Reply:
i.e. I'm wondering about doing this in a loop until the missing NSRTs or the target NSRTs are empty

Reply:
note that this is a recursive function. So it will keep extending the analogy as long as it can find some useful matches.

Reply:
ahh right sorry I totally missed the recurison. makes sense!